File: csharp/rocketmq-client-csharp/PushSubscriptionSettings.cs
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using Google.Protobuf.WellKnownTypes;
using Microsoft.Extensions.Logging;
using Proto = Apache.Rocketmq.V2;

namespace Org.Apache.Rocketmq
{
    public class PushSubscriptionSettings : Settings
    {
        private static readonly ILogger Logger = MqLogManager.CreateLogger<PushSubscriptionSettings>();

        private readonly Resource _group;
        private readonly ConcurrentDictionary<string, FilterExpression> _subscriptionExpressions;
        private volatile bool _fifo = false;
        private volatile int _receiveBatchSize = 32;
        private TimeSpan _longPollingTimeout = TimeSpan.FromSeconds(30);

        public PushSubscriptionSettings(string namespaceName, string clientId, Endpoints endpoints, string consumerGroup,
            TimeSpan requestTimeout, ConcurrentDictionary<string, FilterExpression> subscriptionExpressions)
            : base(namespaceName, clientId, ClientType.PushConsumer, endpoints, requestTimeout)
        {
            _group = new Resource(namespaceName, consumerGroup);
            _subscriptionExpressions = subscriptionExpressions;
        }

        public bool IsFifo()
        {
            return _fifo;
        }

        public int GetReceiveBatchSize()
        {
            return _receiveBatchSize;
        }

        public TimeSpan GetLongPollingTimeout()
        {
            return _longPollingTimeout;
        }

        public override Proto.Settings ToProtobuf()
        {
            var subscriptionEntries = new List<Proto.SubscriptionEntry>();
            foreach (var (key, value) in _subscriptionExpressions)
            {
                var topic = new Proto.Resource()
                {
                    ResourceNamespace = Namespace,
                    Name = key
                };
                var filterExpression = new Proto.FilterExpression()
                {
                    Expression = value.Expression
                };
                switch (value.Type)
                {
                    case ExpressionType.Tag:
                        filterExpression.Type = Proto.FilterType.Tag;
                        break;
                    case ExpressionType.Sql92:
                        filterExpression.Type = Proto.FilterType.Sql;
                        break;
                    default:
                        Logger.LogWarning($"[Bug] Unrecognized filter type={value.Type} for push consumer");
                        break;
                }

                var subscriptionEntry = new Proto.SubscriptionEntry
                {
                    Topic = topic,
                    Expression = filterExpression
                };

                subscriptionEntries.Add(subscriptionEntry);
            }

            var subscription = new Proto.Subscription
            {
                Group = _group.ToProtobuf(),
                Subscriptions = { subscriptionEntries }
            };

            return new Proto.Settings
            {
                AccessPoint = Endpoints.ToProtobuf(),
                ClientType = ClientTypeHelper.ToProtobuf(ClientType),
                RequestTimeout = Duration.FromTimeSpan(RequestTimeout),
                Subscription = subscription,
                UserAgent = UserAgent.Instance.ToProtobuf()
            };
        }

        public override void Sync(Proto.Settings settings)
        {
            if (Proto.Settings.PubSubOneofCase.Subscription != settings.PubSubCase)
            {
                Logger.LogError($"[Bug] Issued settings doesn't match with the client type, clientId={ClientId}, " +
                                $"pubSubCase={settings.PubSubCase}, clientType={ClientType}");
            }

            var subscription = settings.Subscription ?? new Proto.Subscription();
            _fifo = subscription.Fifo;
            _receiveBatchSize = subscription.ReceiveBatchSize;
            _longPollingTimeout = subscription.LongPollingTimeout?.ToTimeSpan() ?? TimeSpan.Zero;
            var backoffPolicy = settings.BackoffPolicy ?? new Proto.RetryPolicy();
            switch (backoffPolicy.StrategyCase)
            {
                case Proto.RetryPolicy.StrategyOneofCase.ExponentialBackoff:
                    RetryPolicy = ExponentialBackoffRetryPolicy.FromProtobuf(backoffPolicy);
                    break;
                case Proto.RetryPolicy.StrategyOneofCase.CustomizedBackoff:
                    RetryPolicy = CustomizedBackoffRetryPolicy.FromProtobuf(backoffPolicy);
                    break;
                default:
                    throw new ArgumentException("Unrecognized backoff policy strategy.");
            }
        }
    }
}

File: php/PushSubscriptionSettings.php
<?php

declare(strict_types=1);

/**
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

namespace Apache\Rocketmq;

use Apache\Rocketmq\V2\ClientType;
use Apache\Rocketmq\V2\Settings as V2Settings;
use Apache\Rocketmq\V2\Subscription;
use Apache\Rocketmq\V2\SubscriptionEntry;
use Apache\Rocketmq\V2\Resource;
use Apache\Rocketmq\V2\FilterExpression as V2FilterExpression;
use Apache\Rocketmq\V2\FilterType;
use Google\Protobuf\Duration;

/**
 * Push subscription settings for PushConsumer.
 *
 * References Java PushSubscriptionSettings:
 * - sync() applies server-returned retry policy
 * - toProtobuf() serializes subscriptions with proper FilterExpression type/expression
 */
class PushSubscriptionSettings extends ClientSettings {

    /** @var string Consumer group */
    private string $consumerGroup;

    /** @var array<string, mixed> Topic => FilterExpression map */
    private array $subscriptionExpressions;

    /** @var int Max cache message count */
    private int $maxCacheMessageCount = 1000;

    /** @var int Max cache message size in bytes */
    private int $maxCacheMessageSize = 67108864; // 64MB

    /** @var int Consumption thread count */
    private int $consumptionThreadCount = 20;

    /** @var Duration|null Long polling timeout */
    private ?Duration $longPollingTimeout = null;

    public function __construct(
        string $namespace,
        string $clientId,
        string $endpoints,
        string $consumerGroup,
        array $subscriptionExpressions = [],
        int $maxAttempts = 3,
        int $requestTimeoutMs = 3000
    ) {
        parent::__construct(
            $namespace,
            $clientId,
            ClientType::PUSH_CONSUMER,
            $endpoints,
            $maxAttempts,
            $requestTimeoutMs
        );

        $this->consumerGroup = $consumerGroup;
        $this->subscriptionExpressions = $subscriptionExpressions;
    }

    /**
     * {@inheritdoc}
     */
    public function toProtobuf(): V2Settings {
        $settings = new V2Settings();
        $settings->setAccessPoint($this->endpoints);
        $settings->setClientType(ClientType::PUSH_CONSUMER);
        $settings->setRequestTimeout($this->requestTimeout);

        $subscription = new Subscription();

        $groupResource = new Resource();
        $groupResource->setName($this->consumerGroup);
        $groupResource->setResourceNamespace($this->namespace);
        $subscription->setGroup($groupResource);

        foreach ($this->subscriptionExpressions as $topic => $filterExpression) {
            $topicResource = new Resource();
            $topicResource->setName($topic);
            $topicResource->setResourceNamespace($this->namespace);

            $entry = new SubscriptionEntry();
            $entry->setTopic($topicResource);

            $v2Filter = new V2FilterExpression();
            if (is_object($filterExpression) && method_exists($filterExpression, 'getExpression')) {
                $v2Filter->setExpression($filterExpression->getExpression());
                if (method_exists($filterExpression, 'getType')) {
                    $type = $filterExpression->getType();
                    $v2Filter->setType($type === 'SQL92' ? FilterType::SQL : FilterType::TAG);
                } else {
                    $v2Filter->setType(FilterType::TAG);
                }
            } elseif (is_string($filterExpression)) {
                $v2Filter->setExpression($filterExpression);
                $v2Filter->setType(FilterType::TAG);
            }

            $entry->setExpression($v2Filter);
            $subscription->getSubscriptions()[] = $entry;
        }

        $settings->setSubscription($subscription);

        if ($this->retryPolicy !== null) {
            $backoffPolicy = new \Apache\Rocketmq\V2\RetryPolicy();
            $backoffPolicy->setMaxAttempts($this->retryPolicy->getMaxAttempts());
            $settings->setBackoffPolicy($backoffPolicy);
        }

        return $settings;
    }

    /**
     * Sync settings from server-returned protobuf.
     *
     * Applies: retryPolicy (maxAttempts, exponentialBackoff).
     */
    public function sync(V2Settings $settings): void {
        if (!$settings->hasSubscription()) {
            Logger::error("[Bug] Unexpected pubSub case in subscription settings sync, clientId={$this->clientId}");
            return;
        }

        // Sync retry policy
        if ($settings->hasBackoffPolicy()) {
            $backoffPolicy = $settings->getBackoffPolicy();
            $maxAttempts = $backoffPolicy->getMaxAttempts();

            if ($backoffPolicy->hasExponentialBackoff()) {
                $exponential = $backoffPolicy->getExponentialBackoff();
                $initialMs = 1000;
                $maxMs = 60000;
                $multiplier = 2.0;

                if ($exponential->hasInitial()) {
                    $initialMs = (int)($exponential->getInitial()->getSeconds() * 1000
                        + $exponential->getInitial()->getNanos() / 1000000);
                }
                if ($exponential->hasMax()) {
                    $maxMs = (int)($exponential->getMax()->getSeconds() * 1000
                        + $exponential->getMax()->getNanos() / 1000000);
                }
                $multiplier = $exponential->getMultiplier() > 0 ? $exponential->getMultiplier() : 2.0;

                $this->retryPolicy = new ExponentialBackoffRetryPolicy(
                    $maxAttempts > 0 ? $maxAttempts : $this->maxAttempts,
                    $initialMs,
                    $maxMs,
                    $multiplier
                );
            }

            if ($maxAttempts > 0) {
                $this->maxAttempts = $maxAttempts;
            }
        }
    }

    public function getConsumerGroup(): string {
        return $this->consumerGroup;
    }

    public function getSubscriptionExpressions(): array {
        return $this->subscriptionExpressions;
    }

    public function addSubscription(string $topic, $filterExpression): self {
        $this->subscriptionExpressions[$topic] = $filterExpression;
        return $this;
    }

    public function getMaxCacheMessageCount(): int {
        return $this->maxCacheMessageCount;
    }

    public function setMaxCacheMessageCount(int $count): self {
        $this->maxCacheMessageCount = $count;
        return $this;
    }

    public function getMaxCacheMessageSize(): int {
        return $this->maxCacheMessageSize;
    }

    public function setMaxCacheMessageSize(int $size): self {
        $this->maxCacheMessageSize = $size;
        return $this;
    }

    public function getConsumptionThreadCount(): int {
        return $this->consumptionThreadCount;
    }

    public function setConsumptionThreadCount(int $count): self {
        $this->consumptionThreadCount = $count;
        return $this;
    }

    public function getLongPollingTimeout(): ?Duration {
        return $this->longPollingTimeout;
    }

    public function setLongPollingTimeout(Duration $timeout): self {
        $this->longPollingTimeout = $timeout;
        return $this;
    }
}
